Send active object IDs to the raylet

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

Part of the stop-gap distributed GC solution.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17353/
Test FAILed.

pcmoritz commented on python/ray/includes/unique_ids.pxi:
```diff
@@ -135,6 +136,17 @@ cdef class ObjectID(BaseID):
         check_id(id)
         self.data = CObjectID.FromBinary(<c_string>id)
+        worker = ray.worker.global_worker
+        # TODO(edoakes): a few ObjectIDs are created before the core worker is
```
Which ones are these? Theoretically it shouldn't be possible to submit tasks or put objects before ray is initialized.

pcmoritz replied:
Ah it might be the object id that warms up the object store. That one should be ok and can get lost (maybe make sure that's indeed the only one and edit the comment to include this insight).

edoakes replied:
Fixed this, it was errantly used in the monitor.

pcmoritz commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -15,7 +16,8 @@ CoreWorker::CoreWorker(
       worker_context_(worker_type, job_id),
-      io_work_(io_service_) {
+      io_work_(io_service_),
+      heartbeat_timer_(io_service_, boost::asio::chrono::milliseconds(100)) {
```
let's make the "100" here (and down below) a configuration argument. I'd also consider bumping it to 500ms and the number of objects to 5000.

edoakes replied:
Changed to take it from the config. I think we should keep it the same as the raylet heartbeat.

edoakes replied:
We're sending at 500ms intervals now.

pcmoritz commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,30 @@ void NodeManager::Heartbeat() {
+    std::vector<ObjectID> active_object_id_vector(active_object_ids.begin(),
+                                                  active_object_ids.end());
+    std::random_shuffle(active_object_id_vector.begin(), active_object_id_vector.end());
+    for (const auto &object_id : active_object_ids) {
```
I think you wanted to iterate over the first max_size entries from active_object_id_vector here, right?

Also there is no need to shuffle everything (there could be a lot of elements in active_object_id_vector), just sampling is sufficient.

edoakes replied:
Oops, not sure how that happened...

As far as I can tell, this is the simplest way to do sampling from a vector prior to C++ 17 - if you have other suggestions let me know. Otherwise if you're very concerned about the performance we can implement our own.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17354/
Test FAILed.

zhijunfu commented:
Thanks, I have a few high level comments.

1. I think it would be very very helpful to have a detailed design doc to list:
   - what the architecture will look like eventually, 
   - what are the steps we plan to take to reach there;
   - for each step, what are the functionalities are needed, what changes are required for the main system components (e.g. core worker, raylet, gcs, plasma ...),  and what interfaces will be added/changed;

A such example design doc can be [Core Worker](https://docs.google.com/document/d/1F1YmjhPghDWRxW9LVsP-nmWCNuhDl3cf0mftRyKGPp4/edit#heading=h.mws3z99djlsc). 

Such a document would help a lot to understand what each individual PR is trying to implement, why this particular approach is chosen, how it will interact with other system components, or whether there's a better way to implement the same functionality. 

2. As @pcmoritz has already mentioned, please make this feature flagged and turned off by default.

ericl commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -93,6 +101,34 @@ CoreWorker::CoreWorker(
+void CoreWorker::SendActiveObjectIDsHeartbeat() {
+  RAY_LOG(DEBUG) << "Sending " << active_object_ids_.size() << " object IDs to raylet.";
+  if (active_object_ids_.size() > RayConfig::instance().raylet_active_object_ids_size()) {
+    RAY_LOG(WARNING) << active_object_ids_.size() << "object IDs are currently in scope. "
```
Won't this be too noisy?

edoakes replied:
This shouldn't ever happen under normal operation - if it does, we probably want a bug report so it's fine to be noisy IMO.

ericl commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -15,7 +16,10 @@ CoreWorker::CoreWorker(
       worker_context_(worker_type, job_id),
-      io_work_(io_service_) {
+      io_work_(io_service_),
+      heartbeat_timer_(io_service_,
```
Can you suppress redundant updates here?

edoakes replied:
Done.

edoakes replied:
Or is your suggestion to suppress updates when the entire set of active IDs is unchanged? That should work and would be simple enough but it's unclear to me if it would have a significant performance impact.

edoakes replied:
It's only a few lines of code and clear enough so I am suppressing them now.

ericl commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,30 @@ void NodeManager::Heartbeat() {
+                                           object_id_bytes.size());
+    }
+  } else {
+    // If there are more than the configured maximum number of object IDs to send per
```
Oh interesting, should it also shuffle on core worker side too?

edoakes replied:
The shuffling is just to sample up to the configured limit. I think it's probably fine to send the non-truncated list from worker->raylet to keep the logic in one place but would be fine sampling in both places.

ericl commented:
You should also run `ray microbenchmark` before and after.this change.

ericl commented on src/ray/core_worker/core_worker.h:
```diff
@@ -94,13 +100,23 @@ class CoreWorker {
+
+  /// Indicates whether or not the active_object_ids map has changed since the
+  /// last time it was sent to the raylet.
+  bool active_object_ids_changed;
```
This is not used?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17358/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17361/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17363/
Test FAILed.

stephanie-wang commented on src/ray/core_worker/core_worker.h:
```diff
@@ -81,8 +81,14 @@ class CoreWorker {
     worker_context_.SetCurrentTaskId(task_id);
   }
 
+  void AddActiveObjectID(const ObjectID &object_id);
```
Doc?

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,30 @@ void NodeManager::Heartbeat() {
     heartbeat_data->add_resource_load_capacity(resource_pair.second);
   }
 
+  std::unordered_set<ObjectID> active_object_ids = worker_pool_.GetActiveObjectIDs();
```
We should also consider the arguments of tasks that are queued in the raylet. Maybe this can be done in a subsequent PR.

edoakes replied:
Good point. I think this will add non-trivial complexity to the PR so let's do it in a follow up.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,30 @@ void NodeManager::Heartbeat() {
+  int64_t max_size = RayConfig::instance().raylet_active_object_ids_size();
+  if (active_object_ids.size() <= max_size) {
+    for (const auto &object_id : active_object_ids) {
+      if (heartbeat_data->active_object_id_size() == max_size) {
```
I don't think this `if` is necessary since we're already in the `if` branch for less than max size.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1284,6 +1316,24 @@ void NodeManager::ProcessNotifyActorResumedFromCheckpoint(const uint8_t *message
+  auto message = flatbuffers::GetRoot<protocol::ActiveObjectIDsHeartbeat>(message_data);
+  std::vector<ObjectID> object_ids = from_flatbuf<ObjectID>(*message->object_ids());
+
+  std::unordered_set<ObjectID> active_object_ids;
```
I think you can initialize with `std::unordered_set<ObjectID> active_object_ids(object_ids.begin(), object_ids.end());`. Might be faster.

stephanie-wang commented on src/ray/raylet/worker_pool.h:
```diff
@@ -121,6 +121,12 @@ class WorkerPool {
   bool HasPendingWorkerForTask(const Language &language, const TaskID &task_id);
 
+  /// Get the set of active object IDs from all workers in the worker pool. If
+  /// there are more than the configured raylet_active_object_ids_size, that
```
I don't think this comment is accurate.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17370/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17372/
Test FAILed.

edoakes commented:
Benchmark results:

Before (commit SHA 01d636247218d8ae361d2d825c2da1bbc041ef7e):
```
single core get calls per second 13361.16 +- 91.08
single core put calls per second 4716.78 +- 11.71
single core put gigabytes per second 11.51 +- 6.2
multi core put calls per second 16768.07 +- 262.34
multi core put gigabytes per second 23.27 +- 0.99
single core tasks sync per second 1646.18 +- 28.8
single core tasks async per second 6494.44 +- 125.88
multi core tasks async per second 7958.15 +- 102.77
single core actor calls sync per second 1223.93 +- 24.15
single core actor calls async per second 1493.14 +- 46.94
multi core actor calls async per second 6413.02 +- 38.12
```

This PR with 100ms heartbeats from worker->raylet:
```
single core get calls per second 13165.16 +- 169.1
single core put calls per second 4686.47 +- 6.91
single core put gigabytes per second 11.78 +- 5.95
multi core put calls per second 15718.82 +- 213.91
multi core put gigabytes per second 23.16 +- 1.09
single core tasks sync per second 1663.26 +- 21.82
single core tasks async per second 6362.32 +- 80.5
multi core tasks async per second 6162.39 +- 360.85
single core actor calls sync per second 1055.08 +- 63.54
single core actor calls async per second 1517.7 +- 22.34
multi core actor calls async per second 5198.94 +- 88.05
```

This PR with 500ms heartbeats from worker->raylet:
```
single core get calls per second 13231.26 +- 100.11
single core put calls per second 4412.26 +- 45.55
single core put gigabytes per second 12.06 +- 6.18
multi core put calls per second 16162.62 +- 260.5
multi core put gigabytes per second 23.18 +- 0.99
single core tasks sync per second 1590.0 +- 11.63
single core tasks async per second 6397.5 +- 142.59
multi core tasks async per second 7061.48 +- 327.98
single core actor calls sync per second 1075.32 +- 96.32
single core actor calls async per second 1468.96 +- 27.16
multi core actor calls async per second 5674.48 +- 97.34
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17375/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17381/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17390/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17398/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17423/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17445/
Test FAILed.

stephanie-wang commented on python/ray/includes/unique_ids.pxi:
```diff
@@ -129,12 +130,26 @@ cdef class UniqueID(BaseID):
+    cdef:
+        CObjectID data
+        object buffer_ref
+        c_bool in_core_worker
```
Can you add a comment saying what this is? I don't really understand what this variable means still after looking at the code below.

stephanie-wang replied:
Is the idea that this should be False for object IDs that don't actually have an associated value?

edoakes replied:
Added a comment.

stephanie-wang commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -108,6 +119,48 @@ CoreWorker::CoreWorker(
+    std::cout << "Sending " << active_object_ids_.size() << " object IDs to raylet."
+              << std::endl;
+    if (active_object_ids_.size() >
+        static_cast<size_t>(RayConfig::instance().raylet_active_object_ids_size())) {
```
Maybe just define `raylet_active_object_ids_size` to be a `size_t` in the config.

stephanie-wang commented on src/ray/common/common_protocol.h:
```diff
@@ -31,6 +31,14 @@ template <typename ID>
+/// @param vector The flatbuffer vector.
+/// @return The unordered set of IDs.
+template <typename ID>
+const std::unordered_set<ID> unordered_set_from_flatbuf(
```
I don't think this is used right now.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -1284,6 +1314,21 @@ void NodeManager::ProcessNotifyActorResumedFromCheckpoint(const uint8_t *message
+  auto message = flatbuffers::GetRoot<protocol::ActiveObjectIDsHeartbeat>(message_data);
+  std::vector<ObjectID> object_ids = from_flatbuf<ObjectID>(*message->object_ids());
+
+  std::unordered_set<ObjectID> active_object_ids(object_ids.begin(), object_ids.end());
```
Do we actually need this to be an `unordered_set` in the backend? It just occurred to me that we always overwrite it anyway.

edoakes replied:
We don't need it for each worker's list, but do need it to deduplicate when combining them for the raylet heartbeat. I am using a set here to keep it consistent and make it clear that this list doesn't contain duplicates.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,28 @@ void NodeManager::Heartbeat() {
+    for (const auto &object_id : active_object_ids) {
+      std::string object_id_bytes = object_id.Binary();
+      heartbeat_data->add_active_object_id(object_id_bytes.data(),
+                                           object_id_bytes.size());
```
I think you can do `to_flatbuf` here.

edoakes replied:
This isn't flatbuf, it's protobuf. Same for other comment.

stephanie-wang commented on src/ray/raylet/node_manager.cc:
```diff
@@ -652,14 +674,19 @@ void NodeManager::HeartbeatAdded(const ClientID &client_id,
   for (const auto &heartbeat_data : heartbeat_batch.batch()) {
+    for (int i = 0; i < heartbeat_data.active_object_id_size(); i++) {
+      active_object_ids.insert(ObjectID::FromBinary(heartbeat_data.active_object_id(i)));
+    }
```
Use `from_flatbuf`?

zhijunfu commented on src/ray/raylet/worker_pool.h:
```diff
@@ -121,6 +121,10 @@ class WorkerPool {
 
+  /// Get the set of active object IDs from all workers in the worker pool.
+  /// \return A set containing the active object IDs.
+  std::unordered_set<ObjectID> GetActiveObjectIDs() const;
```
this set might be big, how about we put it as a reference arg instead of a return value to avoid a copy?

edoakes replied:
This will be optimized away:
https://shaharmike.com/cpp/rvo/

zhijunfu commented on src/ray/raylet/worker.cc:
```diff
@@ -113,6 +113,14 @@ void Worker::AcquireTaskCpuResources(const ResourceIdSet &cpu_resources) {
+}
+
+void Worker::SetActiveObjectIds(std::unordered_set<ObjectID> &object_ids) {
+  active_object_ids_ = object_ids;
```
You could save a copy here by changing it to `std::unordered_set<ObjectID>` or `std::unordered_set<ObjectID> &&`, and change this line to:
`active_object_ids_ = std::move(object_ids);`

The caller can be changed to use `worker. SetActiveObjectIds(std::move(object_ids));`

edoakes replied:
Done.

zhijunfu replied:
this needs to be a non-const && to be able to avoid the copy?

also an explicit std::move() should be needed too.

zhijunfu commented on src/ray/raylet/raylet_client.h:
```diff
@@ -176,6 +176,11 @@ class RayletClient {
+  /// Notifies the raylet of the object IDs currently in use on this worker.
+  /// \param object_ids The set of object IDs currently in use.
+  /// \return ray::Status
+  ray::Status ActiveObjectIDsHeartbeat(const std::unordered_set<ObjectID> &object_ids);
```
How about `ReportActiveObjectIDs` or `NotifyActiveObjectIDs`?

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -326,6 +326,28 @@ void NodeManager::Heartbeat() {
+    }
+  } else {
+    // If there are more than the configured maximum number of object IDs to send per
+    // heartbeat, sample from them randomly.
```
Add a TODO here for further optimizations on the sampling strategy? e.g. prefer objects that have longer last refreshed timestamp

zhijunfu commented on src/ray/raylet/node_manager.cc:
```diff
@@ -652,14 +674,19 @@ void NodeManager::HeartbeatAdded(const ClientID &client_id,
+  std::unordered_set<ObjectID> active_object_ids;
   for (const auto &heartbeat_data : heartbeat_batch.batch()) {
+    for (int i = 0; i < heartbeat_data.active_object_id_size(); i++) {
+      active_object_ids.insert(ObjectID::FromBinary(heartbeat_data.active_object_id(i)));
```
this `active_object_ids` is not used?

zhijunfu replied:
if this is not used in this PR, please add a TODO here?

edoakes replied:
Right, this isn't used yet. Will be used in a follow-up. Adding a TODO.

zhijunfu commented on src/ray/common/ray_config_def.h:
```diff
@@ -20,8 +20,8 @@ RAY_CONFIG(int64_t, ray_cookie, 0x5241590000000000)
 
-/// The duration between heartbeats. These are sent by the raylet.
-RAY_CONFIG(int64_t, heartbeat_timeout_milliseconds, 100)
+/// The duration between heartbeats send from the raylets.
```
```suggestion
/// The duration between heartbeats sent to other raylets.
```

edoakes replied:
This would actually be a bit misleading, because the heartbeats are sent from the raylets to the GCS, not directly to other raylets. Will fix wording though.

zhijunfu commented on src/ray/common/ray_config_def.h:
```diff
@@ -79,6 +83,9 @@ RAY_CONFIG(int64_t, max_num_to_reconstruct, 10000)
 RAY_CONFIG(int64_t, raylet_fetch_request_size, 10000)
 
+/// The maximum number of active object IDs to report in a heartbeat.
+RAY_CONFIG(int64_t, raylet_active_object_ids_size, 1000)
```
```suggestion
RAY_CONFIG(int64_t, raylet_max_active_object_ids, 1000)
```

zhijunfu replied:
`max_active_object_ids` seems more obvious since it's number of object ids, not object sizes.

zhijunfu commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -108,6 +119,48 @@ CoreWorker::CoreWorker(
+  // Only send a heartbeat when the set of active object IDs has changed because the
+  // raylet only modifies the set of IDs when it receives a heartbeat.
+  if (active_object_ids_updated_) {
+    std::cout << "Sending " << active_object_ids_.size() << " object IDs to raylet."
```
change this to RAY_LOG(DEBUG) ?

AmplabJenkins commented:
Can one of the admins verify this patch?

edoakes commented:
@zhijunfu I've addressed all of your comments, please take another look. I also changed the sampling to use reservoir sampling instead of naive sampling.

zhijunfu commented on src/ray/util/sample_test.cc:
```diff
@@ -7,47 +7,51 @@ namespace ray {
-  virtual void TearDown() { delete test_vector; }
+  virtual void TearDown() {
+    delete sample;
+    delete test_vector;
```
not a big deal for UTs, but could use smart pointers to avoid manual deletion.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17594/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17691/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17715/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17721/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17725/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/17746/
Test FAILed.
